src/Microsoft.Health.ControlPlane.CosmosDb/Features/Storage/ControlPlaneDataStore.cs

```diff
@@ -0,0 +1,142 @@
+            EnsureArg.IsNotNull(logger, nameof(logger));
+
+            var collectionConfig = namedCosmosCollectionConfigurationAccessor.Get(Constants.CollectionConfigurationName);
+
+            _documentClient = documentClient;
+            _collectionUri = cosmosDataStoreConfiguration.GetRelativeCollectionUri(collectionConfig.CollectionId);
+            _cosmosDocumentQueryFactory = cosmosDocumentQueryFactory;
+            _logger = logger;
+        }
+
+        public async Task<IdentityProvider> GetIdentityProviderAsync(string name, CancellationToken cancellationToken)
+        {
+            EnsureArg.IsNotNull(name, nameof(name));
+
+            var identityProvider = await GetSystemDocumentByIdAsync<CosmosIdentityProvider>(name, CosmosIdentityProvider.IdentityProviderPartition, cancellationToken);
+            if (identityProvider == null)
+            {
+                throw new IdentityProviderNotFoundException(name);
+            }
+
+            return identityProvider.ToIdentityProvider();
+        }
+
+        public async Task<IdentityProvider> UpsertIdentityProviderAsync(IdentityProvider identityProvider, CancellationToken cancellationToken)
+        {
+            EnsureArg.IsNotNull(identityProvider, nameof(identityProvider));
+
+            var cosmosIdentityProvider = new CosmosIdentityProvider(identityProvider);
+            var resultIdentityProvider = await UpsertSystemObjectAsync(cosmosIdentityProvider, CosmosIdentityProvider.IdentityProviderPartition, cancellationToken);
+            return resultIdentityProvider.ToIdentityProvider();
+        }
+
+        internal IDocumentQuery<T> CreateDocumentQuery<T>(
+            SqlQuerySpec sqlQuerySpec,
+            FeedOptions feedOptions = null)
+        {
+            EnsureArg.IsNotNull(sqlQuerySpec, nameof(sqlQuerySpec));
+
+            var context = new CosmosQueryContext(_collectionUri, sqlQuerySpec, feedOptions);
+
+            return _cosmosDocumentQueryFactory.Create<T>(_documentClient.Value, context);
+        }
+
+        private async Task<T> GetSystemDocumentByIdAsync<T>(string id, string partitionKey, CancellationToken cancellationToken)
+        {
+            EnsureArg.IsNotNull(id, nameof(id));
+            EnsureArg.IsNotNull(partitionKey, nameof(partitionKey));
+            var documentQuery = new SqlQuerySpec(
+                "SELECT * FROM root r WHERE r.id = @id",
+                new SqlParameterCollection(new[] { new SqlParameter("@id", id) }));
+            var feedOptions = new FeedOptions
+            {
+                PartitionKey = new PartitionKey(partitionKey),
+            };
+            IDocumentQuery<T> cosmosDocumentQuery =
+                CreateDocumentQuery<T>(documentQuery, feedOptions);
+            using (cosmosDocumentQuery)
+            {
+                var response = await cosmosDocumentQuery.ExecuteNextAsync(cancellationToken);
```

Comment:
nit: I think we agreed on using var on if the type is obvious. so documentQuery/feedOptions above are fine but this one should be explicitly typed.